core/server/common/src/main/java/alluxio/master/journal/raft/transport/CopycatGrpcBaseConnection.java

```diff
@@ -360,37 +360,37 @@ private String responseObjectType(Object responseObject) {
 
   @Override
   public CompletableFuture<Void> close() {
-    if (!mClosed) {
-      LOG.debug("Closing '{}' connection.", mConnectionOwner);
-
-      // Connection can't be used after this.
-      // Lock and set the state.
-      try (LockResource lock = new LockResource(mStateLock.writeLock())) {
-        mClosed = true;
-      }
+    return CompletableFuture.runAsync(() -> {
+      if (!mClosed) {
+        LOG.debug("Closing '{}' connection.", mConnectionOwner);
+
+        // Connection can't be used after this.
+        // Lock and set the state.
+        try (LockResource lock = new LockResource(mStateLock.writeLock())) {
+          mClosed = true;
+        }
 
-      // Stop timeout timer.
-      mTimeoutScheduler.cancel();
+        // Stop timeout timer.
+        mTimeoutScheduler.cancel();
 
-      // Complete underlying gRPC stream.
-      if (!mStreamCompleted) {
-        try {
-          mTargetObserver.onCompleted();
-        } catch (Exception e) {
-          LOG.debug("Completing underlying gRPC stream failed.", e);
+        // Complete underlying gRPC stream.
+        if (!mStreamCompleted) {
+          try {
+            mTargetObserver.onCompleted();
+          } catch (Exception e) {
+            LOG.debug("Completing underlying gRPC stream failed.", e);
+          }
         }
-      }
 
-      // Close pending requests.
-      failPendingRequests(new ConnectException("Connection closed."));
+        // Close pending requests.
+        failPendingRequests(new ConnectException("Connection closed."));
 
-      // Call close listeners.
-      for (Listener<Connection> listener : mCloseListeners) {
-        listener.accept(this);
+        // Call close listeners.
+        for (Listener<Connection> listener : mCloseListeners) {
+          listener.accept(this);
+        }
```

Comment:
Unrelated to this PR - with the way we close the connection, is it possible that we get lingering requests that do not get failed due to them being submitted after `failPendingRequests`?

Reply:
Early in `close()` a state lock is held exclusively and connection is marked as closed. After that new requests always check for the state and should simply fail with `ConnectionClosed`. So I think we're safe against this particular scenario.

Reply:
Could new requests already be past the state check but not put into the pending requests yet?

Reply:
State lock is held in shared mode by clients for the whole duration of call.